Implement leap motion hand tracking

FejZa opened the pull request:
# XRTK - Mixed Reality Toolkit Pull Request

## Overview

Adds hand tracking support for using the leap motion sensor.
Blocked by https://github.com/XRTK/XRTK-Core/pull/552

FejZa committed:
Import codebase

StephenHodgson reviewed (changes requested):
I really love the enthusiasm, but I think I wanna handle the plugins a bit differently. 
Can I take a stab at a first pass for this?

StephenHodgson commented:
Mainly check out the work I'm doing in the SteamVR and Lumin packages respectively

For the most part we're just copying the files and keeping them in the ignored file list. We we're not checking in any of that stuff directly into our repos

FejZa commented:
This is a work in progress Stephen.

FejZa commented:
I checked those in by accident. Will remove them.

FejZa committed:
Remove SDK components falsly checked in

FejZa commented:
It's cleaned up now :). Only the actually hand related and XRTK code left.

StephenHodgson committed:
Merge branch 'development' into feature/hand-tracking

StephenHodgson committed:
Merge branch 'development' into feature/hand-tracking

FejZa committed:
Update ultraleap module to master branch

FejZa committed:
Merge branch 'development' of https://github.com/XRTK/Ultraleap into feature/hand-tracking

FejZa committed:
Update UnityModules to main branch

FejZa committed:
Refactor data provider to fit new hands flow

FejZa committed:
Allow unsafe code for Ultraleap assembly

error CS0227: Unsafe code may only appear if compiling with /unsafe. Enable "Allow 'unsafe' code" in the inspector for 'Packages/com.xrtk.ultraleap/Runtime/XRTK.Ultraleap.asmdef' to fix this error.

FejZa committed:
Temp remove WSAPlayer config for plugin

WMR\XRTK-Core\Plugins\ARM64\LeapC.dll" because it was not found.

FejZa committed:
Wrap up LeapMotionHandControllerDataProvider

FejZa committed:
Clean up meshing related code

FejZa committed:
Fix right to left-hand-system conversion

FejZa committed:
Fix palm position

FejZa committed:
Fix palm rotation

StephenHodgson committed:
Merge branch 'development' into feature/hand-tracking

FejZa committed:
Restore assembly reference to LeapCSharp and update submodule checkout

FejZa committed:
Formatting

FejZa committed:
Rename operation modes

FejZa committed:
Create platform configurations profile

FejZa committed:
Temp implementation to check for leap service

FejZa committed:
Use Ultraleap instead of leap motion where possible

FejZa committed:
Move coordinate conversion to extension

FejZa committed:
Fix left handed conversion for rotations

FejZa committed:
Fill missing joints

StephenHodgson commented on XRTK.Ultraleap/Packages/com.xrtk.ultraleap/Runtime/Extensions/QuaternionExtensions.cs:
```diff
@@ -14,7 +14,14 @@ public static class QuaternionExtensions
+        /// </summary>
+        /// <param name="vector">The <see cref="LeapInternal.LEAP_QUATERNION"/> to convert.</param>
+        /// <returns>A <see cref="Quaternion"/> with identical values to the input <see cref="LeapInternal.LEAP_QUATERNION"/>.</returns>
+        public static Quaternion ToQuaternion(this LeapInternal.LEAP_QUATERNION q) => new Quaternion(q.x, q.y, q.z, q.w);
```
I think it'd be better to do an implicit cast for these. Check out what I did in the Lumin package. You could update it to implicitly cast from `LeapInternal.LEAP_QUATERNION` to `Quaternion`

FejZa committed:
Clean up converter

FejZa committed:
Fix offset docs

FejZa committed:
Fix hand root pose offset

FejZa committed:
Minor reformat for GetHandOffsetPose

FejZa committed:
Fix inspector still referencing old serialized prop name

FejZa committed:
Adjust desktop mode offset for comfort

FejZa committed:
Move conversion into data provider

FejZa committed:
Fix inspector help text

FejZa committed:
Improve ultraleap docs

FejZa committed:
Fix rotation conversion

FejZa committed:
Fix desktop offset does not take camera rotation into account

FejZa committed:
Fold in playspace rotation

FejZa committed:
Fix missing space in inspector

FejZa committed:
Create standalone hand profiles

FejZa committed:
Introduce UltraleapDeviceOffsetMode config

FejZa committed:
Request HML optimize policy if in headset mounted mode

FejZa committed:
Add device offset properties to data provider

FejZa committed:
Merge branch 'development' of https://github.com/XRTK/Ultraleap into feature/hand-tracking

FejZa committed:
Add missing docs

FejZa committed:
Improve formatting in data provider

FejZa committed:
Use proprties for offset pose

FejZa committed:
Merge branch 'development' into feature/hand-tracking

FejZa committed:
Implement Ultraleap package installer

FejZa committed:
Merge branch 'development' of https://github.com/XRTK/Ultraleap into feature/hand-tracking

# Conflicts:
#	XRTK.Ultraleap/Packages/com.xrtk.ultraleap/Editor/UltraleapPluginUtility.cs

FejZa committed:
Merge remote-tracking branch 'origin/development' into feature/hand-tracking

FejZa committed:
Update to 2014.4 LTS

FejZa committed:
Merge remote-tracking branch 'origin/development' into feature/hand-tracking

# Conflicts:
#	XRTK.Ultraleap/Packages/manifest.json
#	XRTK.Ultraleap/Packages/packages-lock.json

FejZa committed:
Introduce UltraleapFrameOptimizationMode

FejZa committed:
Introduce ReconnectionInterval and MaxReconnectionAttempts

FejZa committed:
Implement UltraleapFrameOptimizationMode

FejZa committed:
Improved and streamlined hand lifecycle

FejZa committed:
Remove unused CurrentFrame, CurrentFixedFrame properties

FejZa committed:
Fix root transform (need to fold in playspace now)

FejZa committed:
Fix teleportation

FejZa committed:
Compute pointer pose (best I can do for now)

StephenHodgson committed:
Merge branch 'development' into feature/hand-tracking

StephenHodgson committed:
updated breaking changes from core

StephenHodgson committed:
removed missing asmdef reference

StephenHodgson committed:
Created new XRTK.Ultraleap.Editor.Inspectors.asmdef

fixed reference to copied LeapMotion.LeapCSharp.asmdef (meta wasn't copied)

StephenHodgson committed:
fixed wsa x64 parsing

StephenHodgson committed:
check if UltraleapPluginUtility is running correctly

StephenHodgson committed:
debug more things

StephenHodgson committed:
fix inspector platform

StephenHodgson commented:
/azp run

StephenHodgson committed:
removed debug logs

StephenHodgson commented:
/azp run

StephenHodgson committed:
updated license

StephenHodgson commented:
/azp run

StephenHodgson commented:
/azp run

StephenHodgson committed:
updated leap motion submodule checkout

updated build pipeline

StephenHodgson committed:
added simple test script

StephenHodgson commented:
@FejZa what's left for this PR?

StephenHodgson committed:
Merge branch 'development' into feature/hand-tracking

# Conflicts:
#	XRTK.Ultraleap/Packages/com.xrtk.ultraleap/README.md
#	XRTK.Ultraleap/Packages/packages-lock.json
#	azure-pipelines.yml

StephenHodgson committed:
Merge branch 'development' into feature/hand-tracking

StephenHodgson committed:
updated project to 2019.4.28

updated core dependencies

StephenHodgson committed:
added assembly info

bumped verson to 0.3.0

StephenHodgson committed:
Removed platform and set the package installer to windows universal and standalone only

StephenHodgson committed:
fixed slashes

StephenHodgson committed:
updated build targets

StephenHodgson committed:
updated ultrealeap submodule

StephenHodgson committed:
only build for standalone windows

StephenHodgson committed:
updated applicatoin identifier

StephenHodgson committed:
replaced plugin utility with submodule reference

StephenHodgson committed:
Revert "replaced plugin utility with submodule reference"

This reverts commit e4e339a40fb4867beb2713c3315d7ba6bb1bca85.

StephenHodgson committed:
added missing assembly infos

StephenHodgson committed:
only create controller in play mode

StephenHodgson reviewed (approved):
Works for me
